File: clients/py/src/features/subinvoke_wrap_wrap.py
from typing import Any, TypedDict
from pydantic import BaseModel, Field, validator
from polywrap_client import PolywrapClient
from polywrap_client_config_builder import PolywrapClientConfigBuilder
from polywrap_wasm import WasmPackage
from pathlib import Path

from validators import UriStr, validate_root_directory


class Args(TypedDict):
    a: int
    b: int


class Wrap(BaseModel):
    directory: Path
    uri: UriStr

    @validator("directory")
    def valid_root_directory(cls, v: Any):
        return validate_root_directory(
            v, Path(__file__).parent.parent.parent.parent.parent
        )


class TestCaseInput(BaseModel):
    root_wrap: Wrap = Field(..., alias="rootWrap")
    sub_wrap: Wrap = Field(..., alias="subWrap")
    method: str
    args: Args


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.parse_obj(input)

    root_wrap_dir = Path(__file__).parent / input_obj.root_wrap.directory
    sub_wrap_dir = Path(__file__).parent / input_obj.sub_wrap.directory

    manifest_root_path = root_wrap_dir / "wrap.info"
    wasm_root_path = root_wrap_dir / "wrap.wasm"

    manifest_sub_path = sub_wrap_dir / "wrap.info"
    wasm_sub_path = sub_wrap_dir / "wrap.wasm"

    with open(manifest_root_path, "rb") as f:
        manifest_root = f.read()

    with open(wasm_root_path, "rb") as f:
        wasm_root_module = f.read()

    with open(manifest_sub_path, "rb") as f:
        manifest_sub = f.read()

    with open(wasm_sub_path, "rb") as f:
        wasm_sub_module = f.read()

    root_wrap_package = WasmPackage(
        file_reader=NotImplemented, manifest=manifest_root, wasm_module=wasm_root_module
    )
    sub_wrap_package = WasmPackage(
        file_reader=NotImplemented, manifest=manifest_sub, wasm_module=wasm_sub_module
    )

    config = (
        PolywrapClientConfigBuilder()
        .set_package(input_obj.root_wrap.uri, root_wrap_package)
        .set_package(input_obj.sub_wrap.uri, sub_wrap_package)
        .build()
    )

    client = PolywrapClient(config)

    print(f"Invoking {input_obj.method}")

    result = client.invoke(
        uri=input_obj.root_wrap.uri,
        method=input_obj.method,
        args=input_obj.args,
    )

    if result:
        print(f"Received: {result}")
        print("Success!")

File: clients/swift/Sources/Readiness/Features/SubinvokeWrapWrap.swift
import PolywrapClient
import Foundation
 
struct SubinvokeWrapWrapTest: Feature {
    func runTestCase(input: Any) throws -> Void {
        guard let inputObj = input as? [AnyHashable: Any] else {
           throw InputError.expectedObject
        }

        let method = try Input.expectString(inputObj["method"]!)

        guard let argsDict = inputObj["args"] as? [AnyHashable: Any],
            let a = argsDict["a"] as? Int,
            let b = argsDict["b"] as? Int else {
            throw InputError.expectedObject
        }
        
        let rootWrap = try verifyWrap(inputObj["rootWrap"]!)
        let subWrap = try verifyWrap(inputObj["subWrap"]!)

        let rootWrapUri = rootWrap.uri
        let rootWrapPackage = try loadWasmPackage(rootWrap.directory)

        let subWrapUri = subWrap.uri
        let subWrapPackage = try loadWasmPackage(subWrap.directory)

        let builder = BuilderConfig()
            .addPackage(rootWrapUri, rootWrapPackage)
            .addPackage(subWrapUri, subWrapPackage)
        
        let client = builder.build()

        print("Invoking \(method)")
        let result: Int? = try? client.invoke(uri: rootWrapUri, method: method, args: AddArgs(a: a, b: b))
        if let result = result {
            print("Received: \(result)")
            print("Success!")
        }
    }

    func verifyWrap(_ wrap: Any) throws -> WrapObj {
        guard let wrapObj = wrap as? [AnyHashable: Any] else {
            throw InputError.expectedObject
        }

        let directory = try Input.expectRootDir(wrapObj["directory"]!, rootDir: "./")
        let uri = try Input.expectUri(wrapObj["uri"]!)
        return WrapObj(directory: directory, uri: uri)
    }

    func loadWasmPackage(_ wrapDir: String) throws -> WasmPackage {
        let reader = ResourceReader(bundle: Bundle.module)
        let module = try reader.readFile(wrapDir)
        return WasmPackage(reader: nil, module: module)
    }
}
